zerver/tests/test_upload.py

```diff
@@ -188,6 +218,59 @@ def test_multiple_claim_attachments(self):
 
         self.assertEquals(Attachment.objects.get(path_id=d1_path_id).messages.count(), 2)
 
+    def test_file_download_authorization(self):
+        subscribed_users = ["hamlet@zulip.com", "iago@zulip.com"]
+        unsubscribed_users = ["othello@zulip.com", "prospero@zulip.com"]
+        for user in subscribed_users:
+            self.subscribe_to_stream(user, "test-subscribe")
+
+        # Make the stream private
+        stream = Stream.objects.get(name='test-subscribe')
+        stream.invite_only = True
+        stream.save()
+
+        self.login("hamlet@zulip.com")
+        fp = StringIO("zulip!")
+        fp.name = "zulip.txt"
+        result = self.client.post("/json/upload_file", {'file': fp})
+        json = ujson.loads(result.content)
+        uri = json["uri"]
+        fp_path_id = re.sub('/user_uploads/', '', uri)
+        body = "First message ...[zulip.txt](http://localhost:9991/user_uploads/" + fp_path_id + ")"
+        self.send_message("hamlet@zulip.com", "test-subscribe", Recipient.STREAM, body, "test")
+        self.client.post('/accounts/logout/')
+
+        # Subscribed user should be able to view file
+        for user in subscribed_users:
+            self.login(user)
+            response = self.client.get(uri)
+            data = "".join(response.streaming_content)
+            self.assertEquals("zulip!", data)
+            self.client.post('/accounts/logout/')
+
+        # Unsubscribed user should not be able to view file
+        for user in unsubscribed_users:
+            self.login(user)
+            response = self.client.get(uri)
+            self.assert_json_error(response, "You are not authorized to view this file.", status_code=403)
+            self.client.post('/accounts/logout/')
+
+        # Make the stream public and the attachment as is_realm_public
+        stream = Stream.objects.get(name='test-subscribe')
+        stream.invite_only = False
+        stream.save()
+        attachment = Attachment.objects.get(path_id=fp_path_id)
+        attachment.is_realm_public = True
+        attachment.save()
+
+        # Now unsubscribed_users should be able to access the files
+        for user in unsubscribed_users:
+            self.login(user)
+            response = self.client.get(uri)
+            data = "".join(response.streaming_content)
+            self.assertEquals("zulip!", data)
+            self.client.post('/accounts/logout/')
```

Comment:
Might make sense to have this second half be its own test where we just use a public stream?